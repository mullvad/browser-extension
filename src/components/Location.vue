<script lang="ts" setup>
import { NButton, NCollapse, NCollapseItem, NSpace } from 'naive-ui';

import IconLabel from '@/components/IconLabel.vue';
import LocationTabs from '@/components/LocationTabs.vue';

import useSocksProxies from '@/composables/useSocksProxies';
import useSocksProxy from '@/composables/useSocksProxy';
import useLocations from '@/composables/useLocations';
import useHistoricConnections from '@/composables/useHistoricConnections/useHistoricConnections';
import getRandomSocksProxy from '@/helpers/getRandomSocksProxy';
import type { HistoricConnection } from '@/composables/useHistoricConnections/HistoricConnections.types';

const { toggleLocations } = useLocations();
const { data: socksProxies, isLoading, isError, error } = useSocksProxies();
const { connectToSocksProxy } = useSocksProxy();
const { storeSocksProxyUsage } = useHistoricConnections();

const clickSocksProxy = (country: string, city: string, hostname: string, port?: number) => {
  storeSocksProxyUsage({ country, city, hostname });
  connectToSocksProxy(hostname, port);
  toggleLocations();
};

const clickCountryOrCity = (country: string, city?: string) => {
  const { hostname, port } = getRandomSocksProxy({
    socksProxies: socksProxies.value,
    country,
    city,
  });
  storeSocksProxyUsage({ country, city });
  connectToSocksProxy(hostname, port);
  toggleLocations();
};

const selectLocation = (connection: HistoricConnection) => {
  const { country, city, hostname } = connection;
  if (hostname) {
    clickSocksProxy(country, city, hostname);
  } else {
    clickCountryOrCity(country, city);
  }
};
</script>

<template>
  <p class="mb-8">
    Select the location where you want to have all your browser traffic routed through.
  </p>
  <p v-if="isLoading" class="text-lg flex items-center">
    <IconLabel text="Loading proxy servers list" type="spinner" />
  </p>
  <p v-else-if="isError">{{ error }}</p>
  <div v-else>
    <LocationTabs :selectLocation="selectLocation" />
    <n-collapse arrow-placement="right">
      <n-collapse-item v-for="{ country, cities } in socksProxies" :key="country" :name="country">
        <template #header>
          <n-button quaternary @click.prevent="clickCountryOrCity(country)">
            {{ country }}
          </n-button>
        </template>
        <n-collapse arrow-placement="right">
          <n-collapse-item
            v-for="{ city, proxyList } in cities"
            :key="city"
            :name="city"
            :title="city"
          >
            <template #header>
              <n-button quaternary @click.prevent="clickCountryOrCity(country, city)">
                {{ city }}
              </n-button>
            </template>
            <n-space vertical>
              <n-button
                v-for="proxy in proxyList"
                :key="proxy.hostname"
                secondary
                medium
                @click="clickSocksProxy(country, city, proxy.hostname, proxy.port)"
              >
                {{ proxy.hostname }}
              </n-button>
            </n-space>
          </n-collapse-item>
        </n-collapse>
      </n-collapse-item>
    </n-collapse>
  </div>
</template>

<style scoped>
p {
  color: var(--light-grey);
}
</style>
