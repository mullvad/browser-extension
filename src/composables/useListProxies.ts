import axios, { AxiosError } from 'axios';
import { useQuery } from 'vue-query';

export type Location = {
  city: string;
  code: string;
  country: string;
  longitude: number;
  latitude: number;
};

export type SocksProxy = {
  online: boolean;
  hostname: string;
  ipv4_address: string;
  ipv6_address: string;
  port: number;
  location: Location;
};

interface Servers {
  [country: string]: {
    [city: string]: SocksProxy[];
  };
}

type City = {
  city: string;
  proxyList: SocksProxy[];
};

export type Country = {
  country: string;
  cities: City[];
};

const groupByCountryAndCity = (data: SocksProxy[]) =>
  data
    .filter((proxy: SocksProxy) => proxy.online)
    .reduce((acc: Servers, proxy: SocksProxy) => {
      const { country, city } = proxy.location;

      // Create country object if not present
      if (!(country in acc)) {
        acc[country] = {};
      }
      // Create a city array if not present
      if (!(city in acc[country])) {
        acc[country][city] = [];
      }
      // Add server to servers
      acc[country][city].push(proxy);
      return acc;
    }, {});

const sortProxiesByCountryAndCity = (grouped: Servers) =>
  Object.entries(grouped)
    .map(([country, cityMap]) => {
      const cities = Object.entries(cityMap)
        .map(([city, proxyList]) => {
          return { city, proxyList } as City;
        })
        .sort(({ city: a }, { city: b }) => a.localeCompare(b));
      return { country, cities } as Country;
    })
    .sort(({ country: a }, { country: b }) => a.localeCompare(b));

const useListProxies = () => {
  const getSocksProxies = async () => {
    const { data } = await axios.get<SocksProxy[]>(
      'https://api.mullvad.net/network/v1-beta1/socks-proxies',
    );
    const grouped = groupByCountryAndCity(data);
    return sortProxiesByCountryAndCity(grouped);
  };

  return useQuery<Country[], AxiosError>('socksProxies', getSocksProxies);
};

export default useListProxies;
