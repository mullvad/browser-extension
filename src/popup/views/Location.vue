<template>
  <div class="location">
    <button @click="handleBack">back</button>
    <details v-for="country in Object.keys(servers)" :key="country" class="country">
      <summary>{{ country }}</summary>

      <details v-for="city in Object.keys(servers[country])" :key="city" class="city">
        <summary>
          {{ city }}
        </summary>

        <div v-for="server in servers[country][city]" :key="server.hostname" class="server">
          <div @click="handleSelect(server.socks_name)">
            {{ server.hostname }}
          </div>
        </div>
      </details>
    </details>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
import router from '../router';

import { Servers } from '@/helpers/servers';
import { localStorage } from '@/helpers/localStorage';
import { getSocksConfig, setSocks } from '@/helpers/socks';

export default Vue.extend({
  name: 'Location',
  data() {
    return {
      servers: {} as Servers,
      currentProtocol: '',
      socksName: '',
    };
  },
  methods: {
    handleSelect(socksName: string) {
      const socksConfig = getSocksConfig(this.currentProtocol, socksName);
      setSocks(true, socksConfig);

      localStorage.socksConfig.set(socksConfig);
      localStorage.socksEnabled.set(true);

      // Redirect to Home route
      router.push('/');
    },
    handleBack() {
      router.push('/');
    },
  },
  async created(): Promise<void> {
    const servers = await localStorage.servers.get();
    const socksProtocols = await localStorage.socksProtocols.get();

    // FIXME: Order servers [1-n]
    this.servers = servers;
    this.currentProtocol = socksProtocols.current;

    // FIXME: Open the view to the current Mullvad server
  },
});
</script>

<style lang="scss">
.country {
  border-bottom: 1px solid #192e45;

  > summary {
    padding: 0.5em;
    list-style: none;

    &:hover {
      cursor: pointer !important;
    }
    // For chrome
    &::-webkit-details-marker {
      display: none;
    }
  }
}

.city {
  padding: 0.5em 0 0.5em 1em;
  background-color: #192e45;

  > summary {
    list-style: none;

    &:hover {
      cursor: pointer !important;
    }
    // For chrome
    &::-webkit-details-marker {
      display: none;
    }
  }

  .server {
    padding: 0.5em 0;

    div {
      padding-left: 1em;
      cursor: pointer !important;

      &:hover {
        background-color: #0c1722;
      }
    }
  }
}
</style>
