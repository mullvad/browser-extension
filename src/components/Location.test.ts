import { mount } from '@vue/test-utils';
import { NCollapseItem } from 'naive-ui';

import Location from '@/components/Location.vue';
import useListProxies from '@/composables/useListProxies';

jest.mock('@/composables/useListProxies', () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock('@/composables/useSocksProxy', () => ({
  __esModule: true,
  default: () => ({ connectToSocksProxy: jest.fn() }),
}));

describe('Location', () => {
  it('should show a loading message', () => {
    (useListProxies as jest.Mock).mockReturnValueOnce({ isLoading: true });
    const wrapper = mount(Location);
    expect(wrapper.text()).toMatch(/loading/i);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should show an error message', () => {
    (useListProxies as jest.Mock).mockReturnValueOnce({ isError: true, error: 'Network error' });
    const wrapper = mount(Location);
    expect(wrapper.text()).toMatch(/network error/i);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should show two countries', () => {
    (useListProxies as jest.Mock).mockReturnValueOnce({
      data: [{ country: 'Australia' }, { country: 'Austria' }],
    });
    const wrapper = mount(Location);

    const countries = wrapper.findAllComponents(NCollapseItem);
    expect(countries).toHaveLength(2);
    expect(countries[0]?.text()).toMatch(/australia/i);
    expect(countries[1]?.text()).toMatch(/austria/i);

    expect(wrapper.element).toMatchSnapshot();
  });
});
