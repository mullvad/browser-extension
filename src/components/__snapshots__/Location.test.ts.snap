// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location should show a loading message 1`] = `
<div
  data-v-app=""
>
  
  <p
    class="mb-8"
  >
     While connected through the proxy, your real location and your VPN location are masked with a private and secure location in the selected region. 
  </p>
  <p
    class="text-lg flex items-center"
  >
    
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
    <span
      class="text-info"
    >
      <i
        class="mr-2 animate-spin n-icon"
        role="img"
        style="--n-bezier: cubic-bezier(.4, 0, .2, 1); font-size: 25px;"
      >
        <svg
          height="1em"
          viewBox="0 0 32 32"
          width="1em"
        >
          <path
            d="M16 3a3 3 0 1 0 .002 6.002A3 3 0 0 0 16 3zM8.937 6.438a2.502 2.502 0 0 0-2.5 2.5c0 1.378 1.122 2.5 2.5 2.5c1.38 0 2.5-1.122 2.5-2.5c0-1.38-1.12-2.5-2.5-2.5zm14.126 1.5c-.551 0-1 .449-1 1c0 .55.449 1 1 1c.55 0 1-.45 1-1c0-.551-.45-1-1-1zM6 13.75a2.25 2.25 0 1 0 .001 4.501A2.25 2.25 0 0 0 6 13.75zm20 1c-.691 0-1.25.559-1.25 1.25s.559 1.25 1.25 1.25s1.25-.559 1.25-1.25s-.559-1.25-1.25-1.25zM8.937 21.063c-1.105 0-2 .894-2 2a1.999 1.999 0 1 0 4 0c0-1.106-.894-2-2-2zm14.126.5a1.5 1.5 0 1 0 0 3a1.5 1.5 0 0 0 0-3zM16 24.25c-.965 0-1.75.785-1.75 1.75s.785 1.75 1.75 1.75s1.75-.785 1.75-1.75s-.785-1.75-1.75-1.75z"
            fill="currentColor"
          />
        </svg>
      </i>
       Loading proxy servers list
    </span>
    <!--v-if-->
    
  </p>
  
</div>
`;

exports[`Location should show an error message 1`] = `
<div
  data-v-app=""
>
  
  <p
    class="mb-8"
  >
     While connected through the proxy, your real location and your VPN location are masked with a private and secure location in the selected region. 
  </p>
  <p>
    Network error
  </p>
  
</div>
`;

exports[`Location should show two countries 1`] = `
<div
  data-v-app=""
>
  
  <p
    class="mb-8"
  >
     While connected through the proxy, your real location and your VPN location are masked with a private and secure location in the selected region. 
  </p>
  <div>
    <!--v-if-->
    <div
      class="n-collapse"
      style="--n-font-size: 14px; --n-bezier: cubic-bezier(.4, 0, .2, 1); --n-text-color: rgb(51, 54, 57); --n-divider-color: rgb(239, 239, 245); --n-title-font-size: 14px; --n-title-text-color: rgb(31, 34, 37); --n-title-font-weight: 400; --n-arrow-color: rgb(51, 54, 57);"
    >
      
      <div
        class="n-collapse-item n-collapse-item--right-arrow-placement"
      >
        <div
          class="n-collapse-item__header"
        >
          <div
            class="n-collapse-item__header-main"
          >
            
            <button
              class="n-button n-button--default-type n-button--medium-type"
              style="--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-bezier-ease-out: cubic-bezier(0, 0, .2, 1); --n-ripple-duration: .6s; --n-opacity-disabled: 0.5; --n-wave-opacity: 0.6; font-weight: 400; --n-color: #0000; --n-color-hover: rgba(46, 51, 56, .09); --n-color-pressed: rgba(46, 51, 56, .13); --n-color-focus: rgba(46, 51, 56, .09); --n-color-disabled: #0000; --n-ripple-color: #0000; --n-text-color: rgb(51, 54, 57); --n-text-color-hover: rgb(51, 54, 57); --n-text-color-pressed: rgb(51, 54, 57); --n-text-color-focus: rgb(51, 54, 57); --n-text-color-disabled: rgb(51, 54, 57); --n-border: 1px solid rgb(224, 224, 230); --n-border-hover: 1px solid #36ad6a; --n-border-pressed: 1px solid #0c7a43; --n-border-focus: 1px solid #36ad6a; --n-border-disabled: 1px solid rgb(224, 224, 230); --n-width: initial; --n-height: 34px; --n-font-size: 14px; --n-padding: 0 14px; --n-icon-size: 18px; --n-icon-margin: 6px; --n-border-radius: 3px;"
              tabindex="0"
              type="button"
            >
              <!---->
              <transition-stub>
                <!---->
              </transition-stub>
              <span
                class="n-button__content"
              >
                Australia
              </span>
              <div
                aria-hidden="true"
                class="n-base-wave"
              />
              <!---->
              <!---->
            </button>
            
            <div
              class="n-collapse-item-arrow"
            >
              <i
                class="n-base-icon"
              >
                <svg
                  fill="none"
                  viewBox="0 0 16 16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5.64645 3.14645C5.45118 3.34171 5.45118 3.65829 5.64645 3.85355L9.79289 8L5.64645 12.1464C5.45118 12.3417 5.45118 12.6583 5.64645 12.8536C5.84171 13.0488 6.15829 13.0488 6.35355 12.8536L10.8536 8.35355C11.0488 8.15829 11.0488 7.84171 10.8536 7.64645L6.35355 3.14645C6.15829 2.95118 5.84171 2.95118 5.64645 3.14645Z"
                    fill="currentColor"
                  />
                </svg>
              </i>
            </div>
            <!---->
          </div>
          <!---->
        </div>
        <transition-stub>
          <!---->
        </transition-stub>
      </div>
      <div
        class="n-collapse-item n-collapse-item--right-arrow-placement"
      >
        <div
          class="n-collapse-item__header"
        >
          <div
            class="n-collapse-item__header-main"
          >
            
            <button
              class="n-button n-button--default-type n-button--medium-type"
              style="--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-bezier-ease-out: cubic-bezier(0, 0, .2, 1); --n-ripple-duration: .6s; --n-opacity-disabled: 0.5; --n-wave-opacity: 0.6; font-weight: 400; --n-color: #0000; --n-color-hover: rgba(46, 51, 56, .09); --n-color-pressed: rgba(46, 51, 56, .13); --n-color-focus: rgba(46, 51, 56, .09); --n-color-disabled: #0000; --n-ripple-color: #0000; --n-text-color: rgb(51, 54, 57); --n-text-color-hover: rgb(51, 54, 57); --n-text-color-pressed: rgb(51, 54, 57); --n-text-color-focus: rgb(51, 54, 57); --n-text-color-disabled: rgb(51, 54, 57); --n-border: 1px solid rgb(224, 224, 230); --n-border-hover: 1px solid #36ad6a; --n-border-pressed: 1px solid #0c7a43; --n-border-focus: 1px solid #36ad6a; --n-border-disabled: 1px solid rgb(224, 224, 230); --n-width: initial; --n-height: 34px; --n-font-size: 14px; --n-padding: 0 14px; --n-icon-size: 18px; --n-icon-margin: 6px; --n-border-radius: 3px;"
              tabindex="0"
              type="button"
            >
              <!---->
              <transition-stub>
                <!---->
              </transition-stub>
              <span
                class="n-button__content"
              >
                Austria
              </span>
              <div
                aria-hidden="true"
                class="n-base-wave"
              />
              <!---->
              <!---->
            </button>
            
            <div
              class="n-collapse-item-arrow"
            >
              <i
                class="n-base-icon"
              >
                <svg
                  fill="none"
                  viewBox="0 0 16 16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5.64645 3.14645C5.45118 3.34171 5.45118 3.65829 5.64645 3.85355L9.79289 8L5.64645 12.1464C5.45118 12.3417 5.45118 12.6583 5.64645 12.8536C5.84171 13.0488 6.15829 13.0488 6.35355 12.8536L10.8536 8.35355C11.0488 8.15829 11.0488 7.84171 10.8536 7.64645L6.35355 3.14645C6.15829 2.95118 5.84171 2.95118 5.64645 3.14645Z"
                    fill="currentColor"
                  />
                </svg>
              </i>
            </div>
            <!---->
          </div>
          <!---->
        </div>
        <transition-stub>
          <!---->
        </transition-stub>
      </div>
      
    </div>
  </div>
  
</div>
`;
